### PR TITLE
add callback support for catalog client, impl ServiceCatalogCache

### PR DESCRIPTION
**Pull request by rkx-bloodshed**

### CODE REVIEW

File: src/main/java/com/orbitz/consul/AgentClient.java [hunk 1]
```diff
@@ -72,7 +72,7 @@ public void ping() {
      * Registers the client as a service with Consul with a ttl check.
      *
      * @param port The public facing port of the service to register with Consul.
-     * @param ttl  Time to live for the Consul dead man's switch.
+     * @param ttl  Time to live in seconds for the Consul dead man's switch.
```
Comment (yfouquet) [inline src/main/java/com/orbitz/consul/AgentClient.java]: This is not related to the review, but good catch

File: src/test/java/com/orbitz/consul/BaseIntegrationTest.java [hunk 2]
```diff
@@ -18,4 +24,16 @@ public static void beforeClass() {
                 .withReadTimeoutMillis(Duration.ofSeconds(2).toMillis())
                 .build();
     }
+
+    @After
```
Comment (yfouquet) [inline src/test/java/com/orbitz/consul/BaseIntegrationTest.java]: This is called after each test; but you do not clean the list.
If test class gets bigger, you will try to deregister services that no longer exists after every test.

File: src/test/java/com/orbitz/consul/BaseIntegrationTest.java [hunk 3]
```diff
@@ -1,12 +1,18 @@
...
 import org.junit.BeforeClass;
 
 import java.time.Duration;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CopyOnWriteArrayList;
 
 public abstract class BaseIntegrationTest {
 
+    private static final List<String> deregisterServices = new CopyOnWriteArrayList<>();
```
Comment (yfouquet) [inline src/test/java/com/orbitz/consul/BaseIntegrationTest.java]: Do not make it static or all tests class will use the same.

File: src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        String serviceId1 = createAutoDeregisterServiceId();
+        String serviceId2 = createAutoDeregisterServiceId();
+
+        List<Map<String, CatalogService>> result = new CopyOnWriteArrayList<>();
+        CountDownLatch finish = new CountDownLatch(1);
+
+        ServiceCatalogCache cache = ServiceCatalogCache.newCache(client.catalogClient(), name);
+        cache.addListener(serviceMap -> {
+            result.add(serviceMap);
+            if (result.size() == 3)
```
Comment (yfouquet) [inline src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java]: why not using CountDownLatch(3)?

File: src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                finish.countDown();
+        });
+
+        cache.start();
+
+        client.agentClient().register(20001, 20, name, serviceId1);
```
Comment (yfouquet) [inline src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java]: if you do not insert a pause between this two lines, the test may become flaky

File: src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        client.agentClient().register(20002, 20, name, serviceId2);
+
+        Uninterruptibles.awaitUninterruptibly(finish, 1, TimeUnit.SECONDS);
```
Comment (yfouquet) [inline src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java]: you could assertTrue on the latch to know if it was actually called 3 times.

File: src/test/java/com/orbitz/consul/CatalogTests.java [hunk 7]
```diff
@@ -182,4 +183,86 @@ public void shouldDeregisterWithDefaultDC() throws InterruptedException {
...
+        List<CatalogService> result = cf.get(1, TimeUnit.SECONDS);
+
+        assertEquals(1, result.size());
+        CatalogService service = result.get(0);
+
+        assertEquals(serviceId, service.getServiceId());
+    }
+
+    @Test
+    public void shouldGetNodeInCallback() throws ExecutionException, InterruptedException, TimeoutException {
```
Comment (yfouquet) [inline src/test/java/com/orbitz/consul/CatalogTests.java]: this is not related to the review (but quite useful), please push it in a later review

Comment (rkx-bloodshed) [reply src/test/java/com/orbitz/consul/CatalogTests.java]: in this PR added method with callback in CatalogClient

File: src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java (hunk 5 above)
Comment (rkx-bloodshed) [reply src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java]: fixed

File: src/test/java/com/orbitz/consul/BaseIntegrationTest.java (hunk 2 above)
Comment (rkx-bloodshed) [reply src/test/java/com/orbitz/consul/BaseIntegrationTest.java]: fixed

**Comment by yfouquet** [comment]
Thanks @rkx-bloodshed

**Pull request merged**